include/seqan3/alignment/matrix/detail/aligned_sequence_builder.hpp

```diff
@@ -0,0 +1,250 @@
+ * into a seqan3::gap_decorator if the expression is not ill-formed, otherwise the transformed aligned sequence type
+ * is a std::vector over the respective alphabet decorated with seqan3::gapped.
+ */
+template <std::ranges::ViewableRange fst_rng_t, std::ranges::ViewableRange sec_rng_t>
+struct aligned_sequence_builder_default_traits
+{
+private:
+    //!\brief Helper type alias for sliced range.
+    using fst_slice_t = decltype(std::declval<fst_rng_t>() | view::slice(0, 1));
+    //!\copydoc aligned_sequence_builder_default_traits::fst_slice_t
+    using sec_slice_t = decltype(std::declval<sec_rng_t>() | view::slice(0, 1));
+
+    //!\brief Helper function to determine the transformed type.
+    template <typename rng_t>
+    //!\cond
+        requires std::Constructible<gap_decorator<rng_t>>
+    //!\endcond
+    static auto aligned_type() { return gap_decorator<rng_t>{}; }
+
+    //!\overload
+    template <typename rng_t>
+    static auto aligned_type() { return std::vector<gapped<value_type_t<rng_t>>>{}; }
+
+public:
+    //!\brief The aligned sequence type for the first sequence.
+    using fst_aligned_t = decltype(aligned_type<fst_slice_t>());
+    //!\brief The aligned sequence type for the second sequence.
+    using sec_aligned_t = decltype(aligned_type<sec_slice_t>());
+};
+
+/*!\brief Builds the alignment for a given pair of sequences and the respective trace.
+ * \ingroup alignment_matrix
+ * \tparam fst_rng_t The first range of the pairwise alignment; must model std::ranges::ViewableRange.
+ * \tparam sec_rng_t The first range of the pairwise alignment; must model std::ranges::ViewableRange.
+ * \tparam traits_t A traits class that determines the seqan3::AlignedSequence type; defaults to
+ *                  seqan3::detail::aligned_sequence_builder_default_traits.
+ *
+ * \details
+ *
+ * This class builds the alignment from a given trace path over the specified sequences. Use the interface
+ * seqan3::detail::aligned_sequence_builder::build_from to get the alignment. The returned seqan3::AlignedSequence type
+ * is determined by the `traits_t` class which needs two member types that represent the seqan::AlignedSequence type.
+ * The specified types must be constructible from the original type refined with a seqan3::view::slice. So the following
+ * expressions must evaluate to `true` in order use the sequences with the given aligned sequence types:
+ * ```cpp
+ * std::Constructible<typename traits_t::fst_aligned_t, decltype(std::declval<fst_rng_t>() | view::slice(0, 1))>
+ * std::Constructible<typename traits_t::sec_aligned_t, decltype(std::declval<sec_rng_t>() | view::slice(0, 1))>
+ * ```
+ */
+template <std::ranges::ViewableRange fst_rng_t,
+          std::ranges::ViewableRange sec_rng_t,
+          typename traits_t = aligned_sequence_builder_default_traits<fst_rng_t, sec_rng_t>>
+class aligned_sequence_builder
```

Comment:
What is the advantage of a class instead of a single free function `build_alignment(seq1, seq2, trace)`?

Reply:
This is an interface strongly coupled with the algorithms and the other classes I added before so there is no need for providing an algorithm like interface. And there will be versions in the future that will maintain a stack so that you can actually parse all possible tracebacks and not just one of the optimal paths as it is right now. So the algorithm is not stateless and therefor the class design is much more flexible.

Reply:
>  And there will be versions in the future that will maintain a stack so that you can actually parse all possible tracebacks and not just one of the optimal paths as it is right now.

Hm I see the use case fo suboptimal paths, but don't you need to stack the trace paths, not the aligned_sequence_builder instances?

> So the algorithm is not stateless

What is the state? because the sequences are passed along anyway within the alignment algorithm (as far as I recall from previous reviews)

> the class design is much more flexible.
 
Can you give me an example where a class design storing the sequences is more flexible than a free function in detail? I am not able to see this as clearly as you do.

Reply:
Well, the sort answer would be try to never use free functions if there is not a real good reason for it (like interface functions that model a concept for types that might not be able to have member functions, or algorithms that are not coupled to the data). Free functions is the way we did things in SeqAn2 out of a design decision. Now we want to have more encapsulation. Something that classes can provide not free functions and will give much more flexibility. In this case the data is strongly coupled with the methods acting on it.  

> Hm I see the use case fo suboptimal paths, but don't you need to stack the trace paths, not the aligned_sequence_builder instances?

Right! And how would you do it? You parse the trace in a DFS order. Now, you can either provide all paths at once because you have an algorithm (since we do not use coroutines yet) or you can do it on demand because you can store the state (which is btw. our general design for SeqAn3 - doing things lazy). Now, I wasn't sure yet if I need a range interface for this but this can be simply added in the future if required. But again, I don't see how this is a function exposed to the public.